src/saltext/vmware/modules/license_mgr.py

```diff
@@ -0,0 +1,207 @@
+from saltext.vmware.utils.connect import get_service_instance
+
+log = logging.getLogger(__name__)
+
+try:
+    from pyVmomi import vim
+
+    HAS_PYVMOMI = True
+except ImportError:
+    HAS_PYVMOMI = False
+
+
+__virtualname__ = "vmware_license_mgr"
+__proxyenabled__ = ["vmware_license_mgr"]
+__func_alias__ = {"list_": "list"}
+
+
+def __virtual__():
+    if not HAS_PYVMOMI:
+        return False, "Unable to import pyVmomi module."
+    return __virtualname__
+
+
+def add(license_key, **kwargs):
+    """
+    Add a license specified by license key to a Datacenter, Cluster, ESXI Server or vCenter
+    If no datacenter, cluster or ESXI Server is specified, it is assumed the operation is to be applied to a vCenter
+
+    license_key
+        License Key which specifies license to add to license manager
+
+    service_instance
+        Use this vCenter service connection instance instead of creating a new one [default None]
+
+    datacenter_name
+        Datacenter name to use for the operation [default None]
+
+    cluster_name
+        Name of the cluster to add license [default None]
+
+    esxi_hostname
+        Hostname of the ESXI Server to add license [default None]
+
+    CLI Example:
+
+    .. code-block: bash
+
+        salt '*' vmware_license_mgr.add license_key datacenter_name=dc1
+    """
+    log.debug("start vmware ext license_mgr add license_key")
+    ret = {}
+    op = {}
+    for key, value in kwargs.items():
+        op[key] = value
+
+    service_instance = op.pop("service_instance", None)
+    datacenter_name = op.pop("datacenter_name", None)
+    cluster_name = op.pop("cluster_name", None)
+    esxi_hostname = op.pop("esxi_hostname", None)
+
+    if service_instance is None:
+        service_instance = get_service_instance(opts=__opts__, pillar=__pillar__)
+
+    if not utils_license_mgr.is_vcenter(service_instance):
+        ret["message"] = "Failed, not connected to a vCenter"
+        ret["result"] = False
+        return ret
+
+    try:
+        if "test" in __opts__ and __opts__["test"]:
+            ret["licenses"] = license_key
+            ret["message"] = "Test dry-run, not really connected to a vCenter testing"
+            return ret
+
+        result = utils_license_mgr.add_license(
+            service_instance, license_key, datacenter_name, cluster_name, esxi_hostname
+        )
+        if result:
+            ret["licenses"] = license_key
+
+    except (
+        salt.exceptions.VMwareApiError,
+        salt.exceptions.VMwareObjectRetrievalError,
+        salt.exceptions.VMwareRuntimeError,
+        salt.exceptions.CommandExecutionError,
+    ) as exc:
+        log.exception(exc)
+        ret[
+            "message"
+        ] = f"Failed to add a license key '{license_key}' due to Exception '{str(exc)}'"
+        ret["result"] = False
+        return ret
+
+    if not result:
+        ret[
+            "message"
+        ] = f"Failed specified license key '{license_key}' was not added to License Manager"
```

Comment:
Given that supply it, don't see the problem since they already know it

Reply:
Mainly because we do that for passwords in Salt - and if it can come from a pillar, the pillar could have GPG encryption. Dumping the license key to a log file may or may not be desirable :shrug:

Reply:
It is passed as a parameter to the function, but I know you mean, provide as little feedback as possible to minimize attack vectors. Will pull it for now, and if we get complaints then can add it back later, noting that the exception could report the license key, but then that is being reported by lower layers, and if it is an issue, they will have to fix what is getting thrown by them.

Reply:
removed license_key from most log.debug messages